frontend/src/main/scala/bloop/testing/TestInternals.scala

```diff
@@ -29,14 +37,83 @@ object TestInternals {
     new FilteredLoader(getClass.getClassLoader, filter)
   }
 
-  @tailrec
-  def executeTasks(tasks: List[TestTask], eventHandler: EventHandler, logger: Logger): Unit = {
-    tasks match {
-      case task :: rest =>
-        val newTasks = task.execute(eventHandler, Array(logger)).toList
-        executeTasks(rest ::: newTasks, eventHandler, logger)
-      case Nil =>
-        ()
+  /**
+   * Execute the test tasks.
+   *
+   * @param processConfig   Configures whether tests should be run in a forked JVM or in process.
+   * @param discoveredTests The tests that were discovered.
+   * @param eventHandler    Handler that reacts on messages from the testing frameworks.
+   * @param logger          Logger receiving test output.
+   */
+  def executeTasks(processConfig: ProcessConfig,
+                   discoveredTests: DiscoveredTests,
+                   eventHandler: EventHandler,
+                   logger: Logger): Unit = {
+    processConfig match {
+      case inProcess: InProcess => executeTasks(inProcess, discoveredTests, eventHandler, logger)
+      case fork: Fork => executeTasks(fork, discoveredTests, eventHandler, logger)
+    }
+  }
+
+  /**
+   * Execute the test tasks in a forked JVM.
+   *
+   * @param fork            Configuration for the forked JVM.
+   * @param discoveredTests The tests that were discovered.
+   * @param eventHandler    Handler that reacts on messages from the testing frameworks.
+   * @param logger          Logger receiving test output.
+   */
+  private def executeTasks(fork: Fork,
+                           discoveredTests: DiscoveredTests,
+                           eventHandler: EventHandler,
+                           logger: Logger): Unit = {
+    logger.debug("Starting forked test execution.")
+
+    val testLoader = fork.toExecutionClassLoader(Some(filteredLoader))
+    val server = new TestServer(logger, eventHandler, discoveredTests)
+    val forkMain = classOf[sbt.ForkMain].getCanonicalName
+    val arguments = Array(server.port.toString)
+    val testAgentFiles = DependencyResolution.resolve(sbtOrg, testAgentId, testAgentVersion, logger)
```

Comment:
It looks like GitHub lost your previous comment... Anyway, I've run a benchmark for that, and the results show that the cost of resolving this dependency with coursier is negligible:

```
[info] Benchmark                                   Mode    Cnt   Score   Error  Units
[info] ResolveBenchmark.resolve                  sample  57525   1.737 ± 0.003  ms/op
[info] ResolveBenchmark.resolve:resolve·p0.00    sample          1.616          ms/op
[info] ResolveBenchmark.resolve:resolve·p0.50    sample          1.698          ms/op
[info] ResolveBenchmark.resolve:resolve·p0.90    sample          1.767          ms/op
[info] ResolveBenchmark.resolve:resolve·p0.95    sample          1.825          ms/op
[info] ResolveBenchmark.resolve:resolve·p0.99    sample          3.074          ms/op
[info] ResolveBenchmark.resolve:resolve·p0.999   sample          4.223          ms/op
[info] ResolveBenchmark.resolve:resolve·p0.9999  sample          7.195          ms/op
[info] ResolveBenchmark.resolve:resolve·p1.00    sample         18.350          ms/op
```

Here's the source of the benchmark:
```scala
package bloop

import java.util.concurrent.TimeUnit

import org.openjdk.jmh.annotations._

@BenchmarkMode(Array(Mode.SampleTime))
@OutputTimeUnit(TimeUnit.MILLISECONDS)
@Fork(value = 1)
@Warmup(iterations = 10, time = 10, timeUnit = TimeUnit.SECONDS)
@Measurement(iterations = 10, time = 10, timeUnit = TimeUnit.SECONDS)
class ResolveBenchmark {

  @Benchmark
  def resolve(): Unit = {
    DependencyResolution.resolve("org.scala-sbt", "test-agent", "1.0.4", logging.NoopLogger)
  }
}
```

Reply:
Okay, if it’s under 20ms is fine for me.